tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/concurrent/TaskWrapper.java

```diff
@@ -0,0 +1,77 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Eclipse RDF4J contributors.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *******************************************************************************/
+package org.eclipse.rdf4j.federated.evaluation.concurrent;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+
+import org.eclipse.rdf4j.federated.FedXConfig;
+
+/**
+ * A {@link TaskWrapper} is a facility to wrap {@link Runnable} background tasks before they are passed to the
+ * {@link Executor}. Use-cases include injection of thread-local context variables, or more fine-granular error
+ * handling.
+ * 
+ * <p>
+ * All sub-queries sent by the federation engine that make use of the concurrency infrastructure (e.g.
+ * {@link ControlledWorkerScheduler}) are passing this wrapper.
+ * </p>
+ * 
+ * <p>
+ * The concrete implementation can be configured using {@link FedXConfig#withTaskWrapper(TaskWrapper)}.
+ * </p>
+ * 
+ * @author Andreas Schwarte
+ * @see ControlledWorkerScheduler
+ * @see FedXConfig#withTaskWrapper(TaskWrapper)
+ */
+public interface TaskWrapper {
+
+	/**
+	 * Default implementation of {@link TaskWrapper} which returns the unmodified original task
+	 */
+	static final TaskWrapper DEFAULT = new TaskWrapper() {
+	};
```

Comment:
Instead of adding a default no-op impl here (which automatically gets inherited by every implementation of TaskWrapper, imho a bit of a code smell) I suggest we instead allow TaskWrapper to be undefined (null or empty) and check for that in relevant places. Also removes the need for default implementations of methods in this interface (which makes discovering what methods to implement easier as well).

Reply:
Done, comment is addressed. Let me know if this works for you